[Serve] Move replica health check to separate concurrency group

### Description

As reported in #24228 and #24419, when a function blocks the event loop, health check times out and replica gets killed. We should move the health check to another asyncio thread/loop to prevent this from happening.  

### Use case

_No response_

Ran a simplified repro and it verifies the hypothesis

```python
def test_healthcheck_timeout(serve_instance):
    @serve.deployment(_health_check_timeout_s=2, _health_check_period_s=1)
    class A:
        def check_health(self):
            return True

        def __call__(self):
            time.sleep(100)

    A.deploy()
    handle = A.get_handle()
    ref = handle.remote()
    ray.get(ref)
```


Gets
```
➜  ~/Desktop/ray/ray/python/ray/serve/tests git:(master) ✗ pytest -v -s test_regression.py -k test_healthcheck_timeout
Test session starts (platform: darwin, Python 3.6.9, pytest 5.3.5, pytest-sugar 0.9.2)
cachedir: .pytest_cache
benchmark: 3.2.2 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/simonmo/Desktop/ray/ray/python
plugins: sugar-0.9.2, rerunfailures-9.1.1, asyncio-0.10.0, custom-exit-code-0.3.0, benchmark-3.2.2, anyio-3.3.4, aiohttp-0.3.0, timeout-1.4.2, lazy-fixture-0.6.3, cov-2.8.1
collecting ... 2022-05-06 11:24:17,510  INFO services.py:1476 -- View the Ray dashboard at http://127.0.0.1:8265
(ServeController pid=58313) INFO 2022-05-06 11:24:24,257 controller 58313 checkpoint_path.py:17 - Using RayInternalKVStore for controller checkpoint and recovery.
(ServeController pid=58313) INFO 2022-05-06 11:24:24,363 controller 58313 http_state.py:115 - Starting HTTP proxy with name 'SERVE_CONTROLLER_ACTOR:SERVE_PROXY_ACTOR-node:127.0.0.1-0' on node 'node:127.0.0.1-0' listening on '127.0.0.1:8000'
(ServeController pid=58313) INFO 2022-05-06 11:24:26,177 controller 58313 deployment_state.py:1217 - Adding 1 replicas to deployment 'A'.
(HTTPProxyActor pid=58318) INFO:     Started server process [58318]
(ServeController pid=58313) INFO 2022-05-06 11:24:31,192 controller 58313 deployment_state.py:498 - Didn't receive health check response for replica A#pUxPRB after 2.0s, marking it unhealthy.
(ServeController pid=58313) INFO 2022-05-06 11:24:33,217 controller 58313 deployment_state.py:498 - Didn't receive health check response for replica A#pUxPRB after 2.0s, marking it unhealthy.
(ServeController pid=58313) INFO 2022-05-06 11:24:35,296 controller 58313 deployment_state.py:498 - Didn't receive health check response for replica A#pUxPRB after 2.0s, marking it unhealthy.
(ServeController pid=58313) INFO 2022-05-06 11:24:35,296 controller 58313 deployment_state.py:563 - Replica A#pUxPRB failed the health check 3times in a row, marking it unhealthy.
(ServeController pid=58313) WARNING 2022-05-06 11:24:35,296 controller 58313 deployment_state.py:1405 - Replica A#pUxPRB of deployment A failed health check, stopping it.


――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_healthcheck_timeout ――――――――――――――――――――――――――――――――――――――――――――――――――――――――

serve_instance = <ray.serve.client.ServeControllerClient object at 0x7fdd08593208>

    def test_healthcheck_timeout(serve_instance):
        @serve.deployment(_health_check_timeout_s=2, _health_check_period_s=1)
        class A:
            def check_health(self):
                return True
    
            def __call__(self):
                time.sleep(100)
    
        A.deploy()
        handle = A.get_handle()
        ref = handle.remote()
>       ray.get(ref)

test_regression.py:251: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../_private/client_mode_hook.py:105: in wrapper
    return func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

object_refs = [ObjectRef(f9e38785c9cda7fb69237282e359f3560e4cab890100000002000000)]

    @PublicAPI
    @client_mode_hook(auto_init=True)
    def get(
        object_refs: Union[ray.ObjectRef, List[ray.ObjectRef]],
        *,
        timeout: Optional[float] = None,
    ) -> Union[Any, List[Any]]:
        """Get a remote object or a list of remote objects from the object store.
    
        This method blocks until the object corresponding to the object ref is
        available in the local object store. If this object is not in the local
        object store, it will be shipped from an object store that has it (once the
        object has been created). If object_refs is a list, then the objects
        corresponding to each object in the list will be returned.
    
        Ordering for an input list of object refs is preserved for each object
        returned. That is, if an object ref to A precedes an object ref to B in the
        input list, then A will precede B in the returned list.
    
        This method will issue a warning if it's running inside async context,
        you can use ``await object_ref`` instead of ``ray.get(object_ref)``. For
        a list of object refs, you can use ``await asyncio.gather(*object_refs)``.
    
        Args:
            object_refs: Object ref of the object to get or a list of object refs
                to get.
            timeout (Optional[float]): The maximum amount of time in seconds to
                wait before returning.
    
        Returns:
            A Python object or a list of Python objects.
    
        Raises:
            GetTimeoutError: A GetTimeoutError is raised if a timeout is set and
                the get takes longer than timeout to return.
            Exception: An exception is raised if the task that created the object
                or that created one of the objects raised an exception.
        """
        worker = global_worker
        worker.check_connected()
    
        if hasattr(worker, "core_worker") and worker.core_worker.current_actor_is_asyncio():
            global blocking_get_inside_async_warned
            if not blocking_get_inside_async_warned:
                logger.warning(
                    "Using blocking ray.get inside async actor. "
                    "This blocks the event loop. Please use `await` "
                    "on object ref with asyncio.gather if you want to "
                    "yield execution to the event loop instead."
                )
                blocking_get_inside_async_warned = True
    
        with profiling.profile("ray.get"):
            is_individual_id = isinstance(object_refs, ray.ObjectRef)
            if is_individual_id:
                object_refs = [object_refs]
    
            if not isinstance(object_refs, list):
                raise ValueError(
                    "'object_refs' must either be an object ref "
                    "or a list of object refs."
                )
    
            # TODO(ujvl): Consider how to allow user to retrieve the ready objects.
            values, debugger_breakpoint = worker.get_objects(object_refs, timeout=timeout)
            for i, value in enumerate(values):
                if isinstance(value, RayError):
                    if isinstance(value, ray.exceptions.ObjectLostError):
                        worker.core_worker.dump_object_store_memory_usage()
                    if isinstance(value, RayTaskError):
                        raise value.as_instanceof_cause()
                    else:
>                       raise value
E                       ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
E                               class_name: create_replica_wrapper.<locals>.RayServeWrappedReplica
E                               actor_id: 69237282e359f3560e4cab8901000000
E                               pid: 58321
E                               name: SERVE_REPLICA::A#pUxPRB
E                               namespace: default_test_namespace
E                               ip: 127.0.0.1
E                       The actor is dead because it was killed by `ray.kill`.

../../worker.py:1841: RayActorError

 ray/serve/tests/test_regression.py::test_healthcheck_timeout ⨯                                                           100% ██████████

Results (21.20s):
       1 failed
         - ray/serve/tests/test_regression.py:239 test_healthcheck_timeout
       8 deselected
```

While the replica shouldn't get killed by health check failure in this case.

--- a/python/ray/serve/common.py
+++ b/python/ray/serve/common.py
@@ -46,6 +46,12 @@ class DeploymentStatusInfo:
         )
 
 
+HEALTH_CHECK_CONCURRENCY_GROUP = "health_check"
+REPLICA_DEFAULT_ACTOR_OPTIONS = {
+    "concurrency_groups": {HEALTH_CHECK_CONCURRENCY_GROUP: 1}
+}
+
+
 class DeploymentInfo:
     def __init__(
         self,
@@ -95,14 +101,14 @@ class DeploymentInfo:
                 or self.serialized_deployment_def is not None
             )
             if self.replica_config.import_path is not None:
-                self._cached_actor_def = ray.remote(
+                self._cached_actor_def = ray.remote(**REPLICA_DEFAULT_ACTOR_OPTIONS)(
                     create_replica_wrapper(
                         self.actor_name,
                         import_path=self.replica_config.import_path,
                     )
                 )
             else:
-                self._cached_actor_def = ray.remote(
+                self._cached_actor_def = ray.remote(**REPLICA_DEFAULT_ACTOR_OPTIONS)(
                     create_replica_wrapper(
                         self.actor_name,
                         serialized_deployment_def=self.serialized_deployment_def,

--- a/python/ray/serve/replica.py
+++ b/python/ray/serve/replica.py
@@ -18,7 +18,7 @@ from ray.util import metrics
 from ray._private.async_compat import sync_to_async
 
 from ray.serve.autoscaling_metrics import start_metrics_pusher
-from ray.serve.common import ReplicaTag
+from ray.serve.common import HEALTH_CHECK_CONCURRENCY_GROUP, ReplicaTag
 from ray.serve.config import DeploymentConfig
 from ray.serve.constants import (
     HEALTH_CHECK_METHOD,
@@ -219,6 +219,7 @@ def create_replica_wrapper(
             if self.replica is not None:
                 return await self.replica.prepare_for_shutdown()
 
+        @ray.method(concurrency_group=HEALTH_CHECK_CONCURRENCY_GROUP)
         async def check_health(self):
             await self.replica.check_health()
 

--- a/python/ray/serve/tests/test_regression.py
+++ b/python/ray/serve/tests/test_regression.py
@@ -238,6 +238,9 @@ def test_uvicorn_duplicate_headers(serve_instance):
 
 def test_healthcheck_timeout(serve_instance):
     # https://github.com/ray-project/ray/issues/24554
+
+    signal = SignalActor.remote()
+
     @serve.deployment(
         _health_check_timeout_s=2,
         _health_check_period_s=1,
@@ -248,13 +251,16 @@ def test_healthcheck_timeout(serve_instance):
             return True
 
         def __call__(self):
-            time.sleep(100)
+            ray.get(signal.wait.remote())
 
     A.deploy()
     handle = A.get_handle()
     ref = handle.remote()
+    # without the proper fix, the ref will fail with actor died error.
     with pytest.raises(GetTimeoutError):
         ray.get(ref, timeout=10)
+    signal.send.remote()
+    ray.get(ref)
 
 
 if __name__ == "__main__":

--- a/python/ray/serve/tests/test_regression.py
+++ b/python/ray/serve/tests/test_regression.py
@@ -1,5 +1,6 @@
 import gc
 import asyncio
+import time
 
 import numpy as np
 import requests
@@ -233,6 +234,27 @@ def test_uvicorn_duplicate_headers(serve_instance):
     resp = requests.get("http://127.0.0.1:8000/A")
     # If the header duplicated, it will be "9, 9"
     assert resp.headers["content-length"] == "9"
+
+
+def test_healthcheck_timeout(serve_instance):
+    # https://github.com/ray-project/ray/issues/24554
+    @serve.deployment(
+        _health_check_timeout_s=2,
+        _health_check_period_s=1,
+        _graceful_shutdown_timeout_s=0,
+    )
+    class A:
+        def check_health(self):
+            return True
+
+        def __call__(self):
+            time.sleep(100)
+
+    A.deploy()
+    handle = A.get_handle()
+    ref = handle.remote()
+    with pytest.raises(GetTimeoutError):
+        ray.get(ref, timeout=10)
 
 
 if __name__ == "__main__":

--- a/python/ray/serve/tests/test_regression.py
+++ b/python/ray/serve/tests/test_regression.py
@@ -1,6 +1,5 @@
 import gc
 import asyncio
-import time
 
 import numpy as np
 import requests

